### DISCUSSION TITLE
:sos: Dealing with STRPTIME('2023-02-14T21:44:21.269380421Z[UTC]') timestamp import/cast :cat2:

### DISCUSSION THREAD

**Discussion by adriens** [duckdb/duckdb#7357] [Q&A] [answered]
# :grey_question: About

i'm currently working on importing an existsing `csv` file into a table, but I cannot manage to finalize my `strptime` call due to parsing issues.
I'm sure that I'm missing someting and wonder if anyone already faced that kind of `string` input for timestamp.

# :pray: `strptime` call

I need to import the following value : 

```
2023-02-14T21:44:20.701374524Z[UTC]
```

into a `TIMESTAMP`.

Here is the current (non-working) `sql` : 

```sql
SELECT strptime('2023-02-14T21:44:21.269380421Z[UTC]', '%Y-%m-%dT%H:%M:%S.%fZ[%Z]') ;
```

Whch actually leads to  : 

<pre>D <font color="#8AE234"><b>SELECT </b></font>STRPTIME(<font color="#C4A000">&apos;2023-02-14T21:44:21.269380421Z[UTC]&apos;</font>, <font color="#C4A000">&apos;%Y-%m-%dT%H:%M:%S.%fZ[UTC]&apos;</font>) <font color="#8AE234"><b>AS </b></font>my_timestamp;
Error: Invalid Input Error: Could not parse string &quot;2023-02-14T21:44:21.269380421Z[UTC]&quot; according to format specifier &quot;%Y-%m-%dT%H:%M:%S.%fZ[UTC]&quot;
2023-02-14T21:44:21.269380421Z[UTC]
                           ^
Error: Literal does not match, expected Z[UTC]
</pre>

Any idea about wha't I'm doing wrong ?

**Comment by adriens** [comment]
Also

<pre>Error: Full specifier did not match: trailing characters
D <font color="#8AE234"><b>SELECT </b></font>strptime(<font color="#C4A000">&apos;2023-02-14T21:44:21.269380421Z&apos;</font>, <font color="#C4A000">&apos;%Y-%m-%dT%H:%M:%S.%fZ&apos;</font>);
Error: Invalid Input Error: Could not parse string &quot;2023-02-14T21:44:21.269380421Z&quot; according to format specifier &quot;%Y-%m-%dT%H:%M:%S.%fZ&quot;
2023-02-14T21:44:21.269380421Z
                           ^
Error: Literal does not match, expected Z
D 
</pre>

**Comment by adriens** [comment]
<pre>D <font color="#8AE234"><b>SELECT </b></font>strptime(<font color="#C4A000">&apos;2023-02-14T21:44:21.269380421&apos;</font>, <font color="#C4A000">&apos;%Y-%m-%dT%H:%M:%S.%f&apos;</font>);
Error: Invalid Input Error: Could not parse string &quot;2023-02-14T21:44:21.269380421&quot; according to format specifier &quot;%Y-%m-%dT%H:%M:%S.%f&quot;
2023-02-14T21:44:21.269380421
                          ^
Error: Full specifier did not match: trailing characters
</pre>

**Comment by Mytherin** [comment, marked as answer]
`%f` matches microseconds which only go up to 6 digits (`999999`). In your case the timestamp has nanosecond precision. This is not yet supported by the `strptime` format specifier but we could add support for that using a different formatting code.

**Reply by adriens** [reply]
Thanks a lot for the very well detailed explanations @Mytherin 

> This is not yet supported by the `strptime` format specifier but we could add support for that using a different formatting code.

In a way I rather that I finished by asking myself what I did not understand in the documentation :sweat_smile: 

:point_right: For now, I'll truncate the data to 6 digits, and will wait for further release.

:bulb: Would you like me to create a dedicated issue :grey_question:

**Reply by adriens** [reply]
Proof : 

<pre><font color="#5FD700">❯</font> duckdb
v0.7.1 b00b93f0b1
Enter &quot;.help&quot; for usage hints.
Connected to a <b>transient in-memory database</b>.
Use &quot;.open FILENAME&quot; to reopen on a persistent database.
D <font color="#8AE234"><b> </b></font>strptime(<font color="#C4A000">&apos;2023-02-14T21:44:21.269380Z[UTC]&apos;</font>, <font color="#C4A000">&apos;%Y-%m-%dT%H:%M:%S.%fZ[%Z]&apos;</font>) ;
┌───────────────────────────────────────────────────────────────────────────┐
│ strptime(&apos;2023-02-14T21:44:21.269380Z[UTC]&apos;, &apos;%Y-%m-%dT%H:%M:%S.%fZ[%Z]&apos;) │
│                         timestamp with time zone                          │
├───────────────────────────────────────────────────────────────────────────┤
│ 2023-02-15 08:44:21.269+11                                                │
└───────────────────────────────────────────────────────────────────────────┘
D 
</pre>
